samza-api/src/main/java/org/apache/samza/operators/descriptors/base/stream/InputDescriptor.java

```diff
@@ -149,6 +155,17 @@ public SubClass withBounded(boolean isBounded) {
     return (SubClass) this;
   }
 
+  /**
+   * Specifies partitions of this stream that all tasks should consume.
+   *
+   * @param broadcastPartitions partitionId or [startingPartitionId-endingPartitionId]. E.g., "0" or "[0-2]"
+   * @return this input descriptor
+   */
+  public SubClass withBroadcastPartitions(String broadcastPartitions) {
```

Comment:
I'm not sure using one of the existing Range classes (apache commons, guava) would be better, since their API is a lot more general. It'll probably be better to write our own simple implementation. Either way, the final validation still happens at runtime (in JavaTaskConfig), since we can't move validation logic from Config classes to Descriptors. Since there's not much validation benefit, and it's fairly clear from documentation what the expected value is, it feels slightly unnecessary to add a new class.

I don't feel strongly about this. Do you prefer one way over the other?

Reply:
I prefer having an explicit class. I agree that it would be better to write our own simple implementation of the class.
An explicit class is a bit more self-documenting and it better prevents accidental mistakes. It also moves validation from run-time to compile-time.
(Less importantly) If we ever want to extend this functionality or remove the config for this in the future, it might be easier to manage with a class.

Reply:
FWIW, it still won't move all validation to compile time. E.g., we can't assert at compile time that lower bound is indeed lower than the upper bound, that none of the bounds are negative, etc. The only thing it will help with is preventing formatting mistakes for the range syntax. 

We could achieve that by changing the method signature to `withBootstrapPartitions(int startIndex, int endIndex)` and document that startIndex <= endIndex. Would that be sufficient?

Reply:
Ah, you are right about that validation. I was thinking about validation for the formatting cases.
For your proposed method, you meant "withBroadcastPartitions", right?
Do you think there will be some other way in the future to specify the broadcast partitions other than with a start/end? For example, one possible pain point is that if an app wants to broadcast all partitions, then the number of partitions needs to be known, and if the number of partitions for a stream changes, then the code needs to change. If we could provide a way to say "all partitions" regardless of partition count, then it might be nice. Another example (less likely) is if someone was a non-continuous list of partitions (e.g. 0,2,4). If you think we might want to add something like these in the future, then it might be better to have a class/interface.